src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -1194,6 +1195,77 @@ export class TerminalService extends Disposable implements ITerminalService {
 	setEditingTerminal(instance: ITerminalInstance | undefined) {
 		this._editingTerminal = instance;
 	}
+
+	createInstanceEventMultiplexer<T>(getEvent: (instance: ITerminalInstance) => Event<T>): { dispose(): void; event: Event<T> } {
+		const store = new DisposableStore();
+		const multiplexer = store.add(new EventMultiplexer<T>());
+		const instanceListeners = store.add(new DisposableMap<ITerminalInstance, IDisposable>());
+
+		function addInstance(instance: ITerminalInstance) {
+			const listener = multiplexer.add(getEvent(instance));
+			instanceListeners.set(instance, listener);
+		}
+
+		// Existing instances
+		for (const instance of this.instances) {
+			addInstance(instance);
+		}
+
+		// Added instances
+		store.add(this.onDidCreateInstance(instance => {
+			addInstance(instance);
+		}));
+
+		// Removed instances
+		store.add(this.onDidDisposeInstance(instance => {
+			instanceListeners.deleteAndDispose(instance);
+		}));
+
+		return {
+			dispose: () => store.dispose(),
+			event: multiplexer.event
+		};
+	}
+
+	createInstanceCapabilityEventMultiplexer<T extends TerminalCapability, K>(capabilityId: T, getEvent: (capability: ITerminalCapabilityImplMap[T]) => Event<K>): { dispose(): void; event: Event<{ instance: ITerminalInstance; data: K }> } {
+		const store = new DisposableStore();
+		const multiplexer = store.add(new EventMultiplexer<{ instance: ITerminalInstance; data: K }>());
+		const capabilityListeners = store.add(new DisposableMap<ITerminalCapabilityImplMap[T], IDisposable>());
+
+		function addCapability(instance: ITerminalInstance, capability: ITerminalCapabilityImplMap[T]) {
+			const listener = multiplexer.add(Event.map(getEvent(capability), data => ({ instance, data })));
+			capabilityListeners.set(capability, listener);
+		}
+
+		// Existing capabilities
```

Comment:
same, you could delete the comments and use variable names that explain the meaning

Reply:
Since it's a bit complicated what's happening I tried to make it extra clear with the comments what's happening